src/Platform/Microsoft.Testing.Extensions.VSTestBridge/ObjectModel/ObjectModelConverters.cs

```diff
@@ -43,84 +82,85 @@ public static TestNode ToTestNode(this TestCase testCase, bool isTrxEnabled, ICl
             {
-                if (property.Id == TestCaseProperties.Id.Id
-                        && getPropertyValue(property) is Guid testCaseId)
-                {
-                    testNode.Properties.Add(new SerializableKeyValuePairStringProperty("vstest.TestCase.Id", testCaseId.ToString()));
-                }
-                else if (property.Id == TestCaseProperties.FullyQualifiedName.Id
-                    && getPropertyValue(property) is string testCaseFqn)
-                {
-                    testNode.Properties.Add(new SerializableKeyValuePairStringProperty("vstest.TestCase.FullyQualifiedName", testCaseFqn));
-                }
-                else if (property.Id == OriginalExecutorUriProperty.Id
-                    && getPropertyValue(property) is Uri originalExecutorUri)
-                {
-                    testNode.Properties.Add(new SerializableKeyValuePairStringProperty("vstest.original-executor-uri", originalExecutorUri.AbsoluteUri));
-                }
+                testNode.Properties.Add(new TestMetadataProperty(category, string.Empty));
+            }
+        }
 
-                // The TP object holding the hierarchy property is defined on adapter utilities and we don't want to enforce that dependency
-                // so instead I use the string ID copied from TP.
-                else if (property.Id == "TestCase.Hierarchy"
-                    && getPropertyValue(property) is string[] testCaseHierarchy
-                    && testCaseHierarchy.Length == 4)
-                {
-                    testNode.Properties.Add(new SerializableNamedArrayStringProperty("vstest.TestCase.Hierarchy", testCaseHierarchy));
-                }
+        if (testCaseOrResult.GetPropertyValue<KeyValuePair<string, string>[]>(TraitsProperty, defaultValue: null) is KeyValuePair<string, string>[] traits &&
+            traits.Length > 0)
+        {
+            foreach (KeyValuePair<string, string> trait in traits)
+            {
+                testNode.Properties.Add(new TestMetadataProperty(trait.Key, trait.Value));
             }
+        }
+    }
 
-            // ID is defined on TraitCollection but is internal so again we copy the string here.
-            if (property.Id == "TestObject.Traits"
-                && getPropertyValue(property) is KeyValuePair<string, string>[] traits && traits.Length > 0)
+    private static void CopyVSTestProviderProperties(IEnumerable<TestProperty> testProperties, TestNode testNode, Func<TestProperty, object?> getPropertyValue)
+    {
+        foreach (TestProperty property in testProperties)
+        {
+            // If vstestProvider is enabled (only known to be true for NUnit and Expecto so far), and we are running server mode in IDE (not dotnet test),
+            // we add these stuff.
+            // Once NUnit and Expecto allow us to move forward and remove vstestProvider, we can remove this logic and get rid of the whole vstestProvider capability.
```

Comment:
It's implicit but I think it would deserve to be made more explicit - you are not relying on client being VS/VSCode but you use the capability because we pass different properties for the client based check.